Fix DX on JavaScripts files and Cypress tests

WiNloSt opened the pull request:
> **Note**
> This is very related to https://github.com/metabase/metabase/pull/28749, I hope we can merge that PR before this one, so I could refactor this PR to only contain a single additional `tsconfig.json` for Cypress directory.

### Motivation
As a VS Code users, I had a lot of painful experience dealing with some JavaScript files and Cypress test files until now. Code navigation is a hit or miss in some JS files it could be hard to find where the functions/components are defined or even being used, and auto import is also not working sometimes.

Also, in Cypress tests, code navigation or autoimport have never worked. As we have a ton of commands living under `e2e/support` it would be ideal to be able to jump right into where those functions are defined.

And this PR attempts to fix that.

### Description
Load both JS and TS files into TypeScript, so code navigation works on JS files as well, and add a new `tsconfig.json` under `e2e` which contains different types for Cypress.

### How to verify
First, test with VS Code, or similar IDE. I've seen Webstorm already works without changing anything, and @nemanjaglumac  told me that IntelliJ also works fine.

1. Try to autoimport or jump to definition in Cypress tests.
2. Try to autoimport or jump to definition in JavaScript files.
3. Try to shift+F12 to see the usage of exported variables in JavaScript files.
4. `yarn type-check` should still work
5. CI checks are green

### Demo
To be added later.

WiNloSt commented on frontend/src/types/cypress.d.ts:
```diff
@@ -0,0 +1,15 @@
+// Fix these types which clashes with Cypress types. This is needed because we're loading Cypress types in some JS files.
```
There's another attempt to achieve the same, but requires a lot of exclusion of files that load Cypress in our main `tsconfig.json` which looks even worse in my opinion. You can see that here. https://github.com/metabase/metabase/pull/28215/commits/ddff71cc8121d1086078d6b6df8012cbaa750065#diff-b55cdbef4907b7045f32cc5360d48d262cca5f94062e353089f189f4460039e0

ranquild replied:
Should we use built-in typescript types instead? https://docs.cypress.io/guides/tooling/typescript-support

WiNloSt replied:
If you mean built-in Cypress types for TypeScript, then I'm already using that.

I use this in any `tsconfig.json` file inside the directories that contain only Cypress file.

```
"types": ["cypress", "@testing-library/cypress"],
```

As far as I understand, this [doc from Cypress](https://docs.cypress.io/guides/tooling/typescript-support), suggests that we create `tsconfig.json` **inside** the cypress folder (which is supposed to contain every cypress-related files, as opposed to us that put both Cypress and non-Cypress files in the same directory).

And this is the problem this file fixes. This file is supposed to be for the main app **not** the cypress file. Since types are already correct by those new `tsconfig.json` files. The problem this file fixes is that since we put both Cypress and non-Cypress file under the same directory e.g. `frontend/test`, by loading the main app files, we're also loading some Cypress files unintentionally, and that causes the Cypress types to get load into TypeScript causing the type conflicts in global names like `expect`, `describe` and whatnot.

Another solution is to exclude every Cypress file in our main app's tsconfig.json which  [was done here](https://github.com/metabase/metabase/commit/ddff71cc8121d1086078d6b6df8012cbaa750065#diff-b55cdbef4907b7045f32cc5360d48d262cca5f94062e353089f189f4460039e0), but IMO looks nastier and probably harder to maintain since the excluded files list might need to be updated.

So, I settled on the middle ground where I left the main app TypeScript also loads the Cypress file (so we will have the type conflicts between jest and Chai), but override Cypress type to point to jest (which we use in our unit test instead).

In an ideal world, I believed the better solution should be moving all cypress-related files under the same directory **and** not putting any non-Cypress-related files inside it e.g. putting unit tests elsewhere, or just creating a new directory called `cypress` and moving all cypress tests there. If we do that, we can remove this `cypress.d.ts` hack + reduce the tsconfig.json files to 2 (1 for the main app, another for Cypress, also another if you count the `tsconfig.base.json`). But that entails we update all the config files e.g. GitHub actions and potentially certain scripts that expect Cypress files to be where they are right now which I don't want to do at the moment. That could be the next milestone, but this improvement is already self-contained, and have an immediate effect on the DX overall. I've been using this myself for the past weeks by merging this PR into any PR I'm working on or reviewing, and it works pretty well for me.

WiNloSt replied:
This isn't needed anymore because https://github.com/metabase/metabase/pull/28749 is merged.

WiNloSt commented on frontend/test/__support__/e2e/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "moduleResolution": "node",
+    "allowJs": true,
+    "paths": {
+      "*": ["../../*", "../../../src/*"]
```
I've only encountered that we import some files from our main app in Cypress-related files in some places, but if we import from any other places that are defined in our main `tsconfig.json` then I'll happily add those, but I just want to keep these configs as concise as possible initially.

https://github.com/metabase/metabase/blob/07c91cd608860798442667074979f514a317b182/tsconfig.json#L26-L35

WiNloSt commented on jsconfig.json:
```diff
@@ -1,21 +0,0 @@
-{
```
jsconfig.json is never used by VS code when we already have tsconfig.json

WiNloSt commented on tsconfig.json:
```diff
@@ -24,14 +24,17 @@
+    "node_modules/",
+    // The following files will load `frontend/src/metabase/app` which will conflict with
+    // `frontend/src/metabase/App.tsx`, since now we're loading both JS and TS files.
+    "frontend/src/metabase/app-main.js",
+    "frontend/src/metabase/app-public.js",
+    "frontend/src/metabase/app-embed.js"
```
without these 3 files excluded, `yarn type-check` would fail.

ranquild reviewed (changes requested):
Have you tried using the approach suggested by the official docs to add typescript support for cypress https://docs.cypress.io/guides/tooling/typescript-support? If yes, could you explain what didn't work so you had to add your own type definitions?

WiNloSt committed:
Attempt to improve DX on code navigation

Previously, JS files isn't picked up by VS Code, so they'll have limited
intellisence. Also, Jest and Cypress has conflicting names e.g.
`describe` or `expect`. So, this commit attempt to resolve that as well.

WiNloSt commented on tsconfig.json:
```diff
@@ -1,38 +1,27 @@
 {
+  "extends": "./tsconfig.base.json",
   "compilerOptions": {
-    "jsx": "react",
```
These options are moved to `tsconfig.base.json` so it can then be reused in both the main app and Cypress' `tsconfig.json`

WiNloSt commented on tsconfig.json:
```diff
@@ -1,38 +1,27 @@
-    "frontend/test/**/*.tsx",
-    "enterprise/frontend/test/**/*.ts",
-    "enterprise/frontend/test/**/*.tsx"
+    "frontend/src/**/*",
```
This is the gist of this PR, loads all the JS files into TypeScript, so intellisense works.

ranquild commented:
@WiNloSt I think it might make sense to wait until @nemanjaglumac finishes his work on moving e2e tests and maybe it would be easier with tsconfig then

ranquild reviewed (approved):
@WiNloSt looks good

nemanjaglumac commented on e2e/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+    "types": ["cypress", "@testing-library/cypress"],
+    "paths": {
+      "e2e/support/*": ["./support/*"],
+      "*": ["../frontend/src/*"]
```
Are you 100% sure we will never need anything from the `.frontend/test/`?

WiNloSt replied:
I'm not 100% sure we **will**. But this could be adjusted when needed. By the way, it doesn't sound that enticing to me if we need to import anything from `frontend/test` in E2E tests, since they're quite different.

nemanjaglumac commented on e2e/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "../tsconfig.base.json",
+  "compilerOptions": {
+    "types": ["cypress", "@testing-library/cypress"],
```
🎉

nemanjaglumac commented on e2e/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+  "compilerOptions": {
+    "types": ["cypress", "@testing-library/cypress"],
+    "paths": {
+      "e2e/support/*": ["./support/*"],
```
Why not `e2e/*`?
Was there a specific reason to scope it down to the support dir only?

WiNloSt replied:
When I need to import stuff from the main app, the following are the result from using VS Code's auto-import.

#### With `"e2e/support/*": ["./support/*"]`
<img width="441" alt="image" src="https://user-images.githubusercontent.com/1937582/222179818-cc3ec17f-824b-4701-9d53-77fa59a0600d.png">

#### WIth `"e2e/*": ["./*"]`
<img width="597" alt="image" src="https://user-images.githubusercontent.com/1937582/222180038-c37ef5c8-e63f-4c29-917d-842209fb17f3.png">

I mean they're actually equivalent, but I'll be cursed by all my teammate.

WiNloSt replied:
Fixed by swapping the position of the `*` and `e2e/*`, so TypeScript tries to resolve the app import using the `*` alias first.

This help not hard-coding `e2e/support/*`, so if we need to refer to any directory under `e2e` in the future, it should work.

nemanjaglumac commented:
@WiNloSt let's backport this, please

WiNloSt commented:
@nemanjaglumac backported
